faculty/session.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2018-2019 ASI Data Science
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from datetime import datetime, timedelta
+from collections import namedtuple
+
+import requests
+import pytz
+from six.moves import urllib
+
+import faculty.config
+
+
+AccessToken = namedtuple("AccessToken", ["token", "expires_at"])
+
+
+class MemoryAccessTokenCache(object):
+    def __init__(self):
+        self._store = {}
+
+    def get(self, profile):
+        access_token = self._store.get(profile)
+        utc_now = datetime.now(tz=pytz.utc)
+        if access_token is None or access_token.expires_at < utc_now:
+            return None
+        else:
+            return access_token
+
+    def add(self, profile, access_token):
+        self._store[profile] = access_token
+
+
+def _service_url(profile, service, endpoint=""):
+    host = "{}.{}".format(service, profile.domain)
+    url_parts = (profile.protocol, host, endpoint, None, None)
+    return urllib.parse.urlunsplit(url_parts)
+
+
+def _get_access_token(profile):
+    url = _service_url(profile, "hudson", "access_token")
+    payload = {
+        "client_id": profile.client_id,
+        "client_secret": profile.client_secret,
+        "grant_type": "client_credentials",
+    }
+
+    response = requests.post(url, json=payload)
+    response.raise_for_status()
+
+    body = response.json()
+
+    token = body["access_token"]
+    now = datetime.now(tz=pytz.utc)
+    expires_at = now + timedelta(seconds=body["expires_in"])
+
+    return AccessToken(token, expires_at)
+
+
+class Session(object):
+
+    __session_cache = {}
+
+    def __init__(self, profile, access_token_cache):
+        self.profile = profile
+        self.access_token_cache = access_token_cache
+
+    @classmethod
+    def get(cls, access_token_cache=None, **kwargs):
+        key = tuple(kwargs.items()) + (access_token_cache,)
```

Comment:
We are not considering the (admittedly unlikely) case where env vars or config files change between invocations of `get`, so that for the same `kwargs` of get you would get different profiles and therefore different sessions. I know there is a cost in calling `resolve_profile`, but it might be worth keying the cache on the resolved profile to make sure we are returning the correct session.

Reply:
Yes, interesting point. @pbugnion do you have any thoughts on this (we had discussed it yesterday).

Reply:
Discussed this with @pbugnion and we decided to stick with the current behaviour.